### DISCUSSION TITLE
Generating wall (solid body) from particles on surface of stl

### DISCUSSION THREAD

**Discussion by Nadun19** [Xiangyu-Hu/SPHinXsys#352] [Q&A] [answered]
I am following the "Example 3: 3D Multi-resolution isotropic particle distribution of complex geometries" to generate SPH particles with an STL file. I was wondering what is the best approach to separate the FluidBody and SolidBody (Wall boundary), which is defined based on the outer surface SPH particles. So the inside SPH particles form the FluidBody. 

Thanks,
Nadun

**Comment by Xiangyu-Hu** [comment]
I am not that I understand you question correctly. Could you specify it with more details?

**Comment by Nadun19** [comment]
Using this example: https://www.sphinxsys.org/html/examples/example3_3D_multiResolutionParticleDistribution.html , I have generated a 3D multi-resolution SPH mesh using an stl file. 

I want to separate these particles into FluidBody and SolidBody. The SolidBody (i.e. a wall) is the layer of particles on the surface of the STL file: 
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/51178787/559d280c-a52b-4308-bb10-d44f371795f8)

and the FluidBody is all the remaining particles inside the geometry.

**Reply by Xiangyu-Hu** [reply]
Now, I understand what you need. However, this is not  along the general practices  of SPH simulation.  We usually need several layers of particles for the solid wall. What is your proposed simulation?

**Reply by Nadun19** [reply]
My intention is to simulate a stomach undergoing peristalsis, therefore the wall will be undergoing a deformation. In this case it is probably essential that there are multiple layers of particles, to stop fluid particles from escaping. What do you think is the best approach for this? I currently only have an STL file of the outside surface (the boundary of the fluid domain).

**Reply by Xiangyu-Hu** [reply]
One way is to generate and relax several layers particles for the stomach wall first. For fluid particles within, you do not need to relaxation particle if you do Lagrangian simulation. You can use the class of exclusive shape.

**Reply by Nadun19** [reply]
All of the sphinxsys examples that generate multiple wall layers uses standard Geometric shapes, but there are no examples on adding several layers to an SPH mesh generated from an stl file. How can I achieve this? My current stls does not have a wall thickness, would I need to generate a new geometry with a wall thickness? or is it possible to extrude/add layers on top of a single layer?

**Reply by Xiangyu-Hu** [reply]
Yes. One way is generate a geometry with a thickness of several layer particles (level set field) from the stl file.
It seems like can be achieved by modify the level set field generated from the original stl.

**Reply by Xiangyu-Hu** [reply]
Another way, probably simpler, 
is define a shape from the original shape, say, from the STL file, using mapping function to obtain one shape with the thickness.
I can set small case today to illustrate the function.

**Comment by Xiangyu-Hu** [comment]
[stomach.pdf](https://github.com/Xiangyu-Hu/SPHinXsys/files/11941030/stomach.pdf)
Here is particle generation on stomach we have done recently.

**Comment by Xiangyu-Hu** [comment, marked as answer]
I have created a issue for the discussion. 
[Xiangyu-Hu](https://github.com/Xiangyu-Hu) Created issue https://github.com/Xiangyu-Hu/SPHinXsys/issues/361
and also created a pull request for the new functionality. #362 
You can check the new example using such functionality, and review the change this pull request. 
Thanks.

**Reply by Nadun19** [reply]
Thanks for this!
